Rosettacode grayscale implementation

budi-kurniawan opened the pull request:
An implementation of Rosettacode's grayscale task (http://rosettacode.org/wiki/Grayscale_image) with additional features, including creating a Bitmap from scratch and manipulating its individual pixels.

privat commented:
Jenkins: ok to test
Jenkins: add to whitelist

privat commented:
Nice proposal since it highlight the need of #1267 .

There is some issues in the code, especially missing documentation on classes and methods, but the documentation on the module is good.

budi-kurniawan commented:
Sure, I'll add documentation to the class and methods.

Thanks,
budi

On Tue, Apr 21, 2015 at 12:52 AM, Jean Privat <notifications@github.com>
wrote:

> Nice proposal since it highlight the need of #1267
> <https://github.com/privat/nit/issues/1267> .
>
> There is some issues in the code, especially missing documentation on
> classes and methods, but the documentation on the module is good.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1282#issuecomment-94638278>.
>

budi-kurniawan commented:
I added comments to my code and committed and pushed it to GitHub.
What is the procedure here, do I need to create another Pull Request? Or,
how do I add the latest commit to the previous Pull Request?

Thanks,
budi

On Tue, Apr 21, 2015 at 12:52 AM, Jean Privat <notifications@github.com>
wrote:

> Nice proposal since it highlight the need of #1267
> <https://github.com/privat/nit/issues/1267> .
>
> There is some issues in the code, especially missing documentation on
> classes and methods, but the documentation on the module is good.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1282#issuecomment-94638278>.
>

Morriar commented:
Nop, as you can see [here](https://github.com/privat/nit/pull/1282/commits), if you push on the same branch it automatically update your PR.

However, you should rebase your branch to merge the two commits. More information about this can be found on the nitlanguage.org website: http://nitlanguage.org/internal/workflow/

budi-kurniawan commented:
Not sure if I did it correctly.

I did "git rebase -i HEAD~1" on my local
Then I did "git push" but this was rejected, so I did a pull, performed
another rebase and pushed again.

Now the PR page looks like this:

https://github.com/privat/nit/pull/1282/commits

On Tue, Apr 21, 2015 at 5:53 PM, Alexandre Terrasa <notifications@github.com
> wrote:

> Nop, as you can see here <https://github.com/privat/nit/pull/1282/commits>,
> if you push on the same branch it automatically update your PR.
>
> However, you should rebase your branch to merge the two commits. More
> information about this can be found on the nitlanguage.org website:
> http://nitlanguage.org/internal/workflow/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1282#issuecomment-94954317>.
>

Morriar commented:
When you was on the interactive rebase, did you merge the commits with the `fixup` or `squash` command?

Also, GitHub always reject fast-forward modification. When rebasing a branch, use the `-f` option to force the fast forward push.

budi-kurniawan commented:
These were the steps:
1. git rebase -i HEAD~2
It then opens a text editor that shows two commits, like this:

pick 01d1124 Rosetta 1
pick 6340aaa Rosetta 2


I then changed the second pick to squash and save the file.

2. git push

rejected.

What's the best way to fix this? Should I just delete the PR and start over?

Morriar commented:
If you just want to fix a commit, you can use `fixup` instead of `squash`. It does the same things but discard the second commit message. This is what you want to do in this case.

To avoid the rejected problem use `git push <repo> <branch> -f`

privat commented on examples/rosettacode/bitmap.nit:
```diff
@@ -0,0 +1,247 @@
+# The Bitmap class represents a 24-bit bitmap image. An instance can be constructed
```
Please add a licence header (look at other nit files)

privat commented on examples/rosettacode/bitmap.nit:
```diff
@@ -0,0 +1,247 @@
+#
+# For information on the bitmap format, see
+# 	http://en.wikipedia.org/wiki/BMP_file_format
+module bitmap
```
In fact, I suggest that you just move the bitmap module from here to lib/, so that other people can more easily discover and use it.

privat commented on examples/rosettacode/bitmap.nit:
```diff
@@ -0,0 +1,247 @@
+# 	http://en.wikipedia.org/wiki/BMP_file_format
+module bitmap
+
+class Bitmap
```
not doc for the class. Maybe part of the for of the module should be here instead.
Or at least add a short doc

privat commented on examples/rosettacode/bitmap.nit:
```diff
@@ -0,0 +1,247 @@
+	var height: Int is noinit
+	private var data_offset: Int is noinit
+	private var bits_per_pixel: Int is noinit
+	private var image_size: Int is noinit
```
Add some doc for the attributes

privat commented on examples/rosettacode/bitmap_test.nit:
```diff
@@ -0,0 +1,43 @@
+# This file tests the Bitmap class in the bitmap module
```
licence header. see other .nit files in the directory

privat commented on examples/rosettacode/bitmap_test.nit:
```diff
@@ -0,0 +1,43 @@
+	end 
+end
+# Save it to blank.bmp
+bitmap2.save("./blank.bmp")
```
I do not know what kind of text editor you are using but it forgot to add a newline character after the last line

privat commented:
good.

For the missing bits, check the list on http://nitlanguage.org/internal/patches/

* license header
* signed-off by
* whitespaces

privat commented:
@budi-kurniawan can you reroll the commits and make them nice an tidy?

budi-kurniawan commented:
Done. Only one nice commit now.

Thanks to Alexandre who's given me a pointer

On Thu, Apr 23, 2015 at 11:28 AM, Jean Privat <notifications@github.com>
wrote:

> @budi-kurniawan <https://github.com/budi-kurniawan> can you reroll the
> commits and make them nice an tidy?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1282#issuecomment-95625287>.
>

privat commented:
jenkins: OK to test

privat commented:
jenkins test this please

privat commented:
still issues with:

* the signed-off-by that does not match the author
* whitespace errors

budi-kurniawan commented:
I've removed the whitespaces. They were not removed previously probably
because I was using the --online option, which was not recognized.
I've also made sure the commit has only one sign-off-by tag.


On Wed, May 6, 2015 at 11:15 AM, Jean Privat <notifications@github.com>
wrote:

> still issues with:
>
>    - the signed-off-by that does not match the author
>    - whitespace errors
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1282#issuecomment-99507643>.
>

privat commented:
The author field is bad. it is `budi2020 <budi2020@ubuntu14>` but something like `Budi Kurniawan <budi2020@gmail.com>` would have been better.
I suspect that you did not configure your git correctly.

budi-kurniawan commented:
sorry, still learning git.
I just modified the author field of my last commit.

On Thu, May 7, 2015 at 8:51 PM, Jean Privat <notifications@github.com>
wrote:

> The author field is bad. it is budi2020 <budi2020@ubuntu14> but something
> like Budi Kurniawan <budi2020@gmail.com> would have been better.
> I suspect that you did not configure your git correctly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1282#issuecomment-100056852>.
>

privat commented:
ok to test

xymus commented on lib/image/bitmap.nit:
```diff
@@ -0,0 +1,259 @@
+#
+# For information on the bitmap format, see
+# 	http://en.wikipedia.org/wiki/BMP_file_format
+
```
The `#` is missing here. The doc is not currently associated to the module.

xymus commented on lib/image/test_bitmap.nit:
```diff
@@ -0,0 +1,44 @@
+# It converts the input bitmap to grayscale and save it to the specified
+# output file. In addition, it creates a blank.bmp file in the current directory.
+#
+
```
Same `#` missing here. Or is this intentionnal?

xymus commented:
This library should be moved to `lib/bitmap`. `image` is too generic for a limited library in my opinion.

xymus commented on lib/image/bitmap.nit:
```diff
@@ -0,0 +1,259 @@
+	end
+
+	# Sets an individual pixel, where position (0, 0) represents the top-left pixel.
+	fun set_pixel(x: Int, y: Int, color: Int)
```
You should use `fun []=(x, y, color: Int)` here to be more nity. The usage would then be `bmp[x, y] = color`.

privat commented:
agree. `bitmap` is a better name than `image` for such a module.

budi-kurniawan commented:
I moved the classes to lib/bitmap. The thinking was that it could be part
of a larger image library, but I guess if that really happens we can always
move them to a lib/image.
In addition, I added the missing # between two comment lines.
I also liked the idea of making set_pixel more nit-ty as per Alexis's
suggestion, but then I thought a bitmap might have a color table and in
such an event, instead of setting a pixel, you would edit the color table.
In this case, set_pixel and edit_color_table would be equally important and
changing set_pixel to a default attribute would leave the impression that
set_pixel is more important.

On Fri, May 8, 2015 at 4:04 PM, Jean Privat <notifications@github.com>
wrote:

> agree. bitmap is a better name than image for such a module.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1282#issuecomment-100343764>.
>

privat commented on lib/bitmap/bitmap.nit:
```diff
@@ -0,0 +1,259 @@
+#	save(path_to_a_file)
+#
+# For information on the bitmap format, see
+#	http://en.wikipedia.org/wiki/BMP_file_format
```
with 4 spaces, the line is condidered a block of Nit code, not a link

privat commented on lib/bitmap/bitmap.nit:
```diff
@@ -0,0 +1,259 @@
+			0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+			0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+
+	# the image data, each cell contains a pixel. The first 8 least-significant bits represent
```
Capitalize `The` and insert a paragraph break after the short description.

privat commented on lib/bitmap/bitmap.nit:
```diff
@@ -0,0 +1,259 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```
Remove the `#

privat commented on lib/bitmap/test_bitmap.nit:
```diff
@@ -0,0 +1,44 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```
Same. remove the `#`

privat commented on lib/bitmap/test_bitmap.nit:
```diff
@@ -0,0 +1,44 @@
+# It has to be called by passing two arguments:
+#
+#	- An input bitmap file
+#	- An output bitmap file
```
I am not sure if `<tab>-` is seend as a list item in markdown. try to use 3 spaces instead.

budi-kurniawan replied:
I've implemented all the required changes.

On Mon, May 11, 2015 at 10:12 AM, Jean Privat <notifications@github.com>
wrote:

> In lib/bitmap/test_bitmap.nit
> <https://github.com/privat/nit/pull/1282#discussion_r30041302>:
>
> > +# you may not use this file except in compliance with the License.
> > +# You may obtain a copy of the License at
> > +#
> > +#     http://www.apache.org/licenses/LICENSE-2.0
> > +#
> > +# Unless required by applicable law or agreed to in writing, software
> > +# distributed under the License is distributed on an "AS IS" BASIS,
> > +# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > +# See the License for the specific language governing permissions and
> > +# limitations under the License.
> > +#
> > +# This file tests the Bitmap class in the bitmap module
> > +# It has to be called by passing two arguments:
> > +#
> > +#	- An input bitmap file
> > +#	- An output bitmap file
>
> I am not sure if <tab>- is seend as a list item in markdown. try to use 3
> spaces instead.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1282/files#r30041302>.
>

budi-kurniawan commented:
I've implemented all the required changes
